File: code/Java/Game.java
// Abstraction and Encapsulation
import java.util.Scanner;

public abstract class Game {
    protected int numberOfGuesses = 0;
    protected int totalscore = 0;

    public void startGame() { // Single player mode
        int targetNumber = generateTargetNumber();
        while (true) {
            int guessedNumber = guessNumber();
            if (guessedNumber > targetNumber) {
                System.out.println("COMPUTER: <Too high>");
            } else if (guessedNumber < targetNumber) {
                System.out.println("COMPUTER: <Too low>");
            } else {
                System.out.print("\033[H\033[2J");//clear output
                System.out.println("\n+++++++++++++ GAME COMPLETED! ++++++++++++++");
                System.out.println("\nCongratulations, my friend!");
                break;
            }
            numberOfGuesses++;
        }
        System.out.println("You guessed it in about " + numberOfGuesses + " attempt!");
    }

    public String startGame2(String player1, String player2) { // multiplayer mode
        int targetNumber = generateTargetNumber();
        Scanner sn = new Scanner(System.in);
        while (true) {
            // PLAYER 1:
            System.out.print("\n" + player1 + ", guess the number: ");
            int guessedNumber1 = sn.nextInt();
            if (guessedNumber1 > targetNumber) {
                System.out.println("COMPUTER: <Too high>");
            } else if (guessedNumber1 < targetNumber) {
                System.out.println("COMPUTER: <Too low>");
            } else {
                System.out.print("\033[H\033[2J");//clear output
                System.out.println("\n+++++++++++++ GAME COMPLETED! ++++++++++++++");
                System.out.println("\nCongratulations, " + player1 + "!");
                System.out.println("<10 points for player 1!>");
                return player1;
            }

            // PLAYER 2:
            System.out.print("\n" + player2 + ", guess the number: ");
            int guessedNumber2 = sn.nextInt();
            if (guessedNumber2 > targetNumber) {
                System.out.println("COMPUTER: <Too high>");
            } else if (guessedNumber2 < targetNumber) {
                System.out.println("COMPUTER: <Too low>");
            } else {
                System.out.print("\033[H\033[2J");//clear output
                System.out.println("\n+++++++++++++ GAME COMPLETED! ++++++++++++++");
                System.out.println("\nCongratulations, " + player2 + "!");
                System.out.println("<10 points for player 2!>");
                return player2;
            }
        }
    }

    protected abstract int generateTargetNumber();

    protected abstract int guessNumber();

    public int getNumberOfGuesses() {
        return numberOfGuesses;
    }

    public int getTotalScore() {
        return totalscore;
    }
}

File: code/C/Game.c
#include "Game.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

void startGame(Game* game) {
    int targetNumber = game->generateTargetNumber(game);
    while (1) {
        int guessedNumber = game->guessNumber(game);
        if (guessedNumber > targetNumber) {
            printf("COMPUTER: <Too high>\n");
        } else if (guessedNumber < targetNumber) {
            printf("COMPUTER: <Too low>\n");
        } else {
            printf("\033[H\033[2J");  // clear output
            printf("\n+++++++++++++ GAME COMPLETED! ++++++++++++++\n");
            printf("\nCongratulations, my friend!\n");
            break;
        }
        game->numberOfGuesses++;
    }
    printf("You guessed it in about %d attempts!\n", game->numberOfGuesses);
}

char* startGame2(Game* game, const char* player1, const char* player2) {
    int targetNumber = game->generateTargetNumber(game);
    char* winner = (char*)malloc(50 * sizeof(char));
    while (1) {
        int guessedNumber1, guessedNumber2;

        // PLAYER 1:
        printf("\n%s, guess the number: ", player1);
        scanf("%d", &guessedNumber1);
        if (guessedNumber1 > targetNumber) {
            printf("COMPUTER: <Too high>\n");
        } else if (guessedNumber1 < targetNumber) {
            printf("COMPUTER: <Too low>\n");
        } else {
            printf("\033[H\033[2J");  // clear output
            printf("\n+++++++++++++ GAME COMPLETED! ++++++++++++++\n");
            printf("\nCongratulations, %s!\n", player1);
            printf("<10 points for player 1!>\n");
            strcpy(winner, player1);
            return winner;
        }

        // PLAYER 2:
        printf("\n%s, guess the number: ", player2);
        scanf("%d", &guessedNumber2);
        if (guessedNumber2 > targetNumber) {
            printf("COMPUTER: <Too high>\n");
        } else if (guessedNumber2 < targetNumber) {
            printf("COMPUTER: <Too low>\n");
        } else {
            printf("\033[H\033[2J");  // clear output
            printf("\n+++++++++++++ GAME COMPLETED! ++++++++++++++\n");
            printf("\nCongratulations, %s!\n", player2);
            printf("<10 points for player 2!>\n");
            strcpy(winner, player2);
            return winner;
        }
    }
}
